# HW 7: Gene Drives Revisited, Dictionaries, and Chat-CMC

## Reading this week 🐶

Please read Chapter 9 of the SCI 10 Coding Companion on the topic of dictionaries.


## Societal Considerations of Gene Drives

Last week, we explored the use of gene drives to combat certain kinds of diseases.

Now, we'd like for you to engage with some current work on CRISPR-based gene drives in Southern California. In particular, this is about a proposal to use gene drives to combat mosquito-borne diseases in Southern California.

Start by watching these three presentations that were made to help educate SoCal residents about the possible use of gene drives.  You may wish to watch them at 1.5 speed. (In total, this should take around 15 minutes.)
  * Video 1: https://tinyurl.com/blossslides1
  * Video 2: https://tinyurl.com/blossslides2
  * Video 3: https://tinyurl.com/blossslides3

**Part 1**

  * What three mosquitio-borne diseases do the videos mention as having been introduced to California?

  <font color="SteelBlue">**Zika, Dengue, Yellow Fever.**</font>  

  * The second video describes two different genetic engineering technologies. The first is GE Sterile Males and the second is Gene Drives. What is the major advantage described _in that video_ of using Gene Drives over GE Sterile Males? (There may be other advantages not mentioned in that video, but we're eager to probe what they discuss in that video.)

  <font color="SteelBlue">**Authorities would only have to release the mosquitoes with gene drives once.**</font>

  * The third video discusses two different types gene drives that could be used to combat mosquito-borne diseases. What are those two types of gene drives? In just one or two sentences describe a major difference between those two approaches.

  <font color="SteelBlue">**The first type of gene drive is aimed towards eliminating the mosquito population as a whole, whereas the second creates mosquitoes that are resistant to the disease.**</font>



**Part 2 (10 Points)**

Based on what you've learned about gene drives for mosquito-borne diseases from those three videos (Part 1), write a memo (_about one or two paragraphs_) to the producers of the videos with recommendations of specific additional information that you believe should be addressed to help SoCal citizens make informed decisions about the possible use of gene drives to combat diseases carried by the _Ae. aegypti_ mosquito. Your recommendation should:

  * Be the actual word-for-word script of the content that should be included in the video.
  * Be new content that is not already covered in those videos.
  * Specify which video (1, 2, or 3) and time stamp in that video where the material should be included so that the content flows most naturally in the current narrative.

You're welcome to use ideas that we learned in class, research that you do on your own, and/or interact with a generative AI tool on this task. However, you're responsible for the content and correctness, so ultimately be sure that what you contribute is something that you can explain and justify.

<font color="SteelBlue">**To be added at 2:46 in Video 3:
We've looked at two different ways gene drive could address the mosquito problem: elimination and modification. Each approach solves different problems, but each comes with trade-offs.
The elimination strategy would create only male offspring, eventually removing Aedes aegypti from Southern California entirely. The benefits are clear: no more dengue, no more bites. But there's an ecological risk. Aedes aegypti is invasive, and removing it completely leaves behind an empty ecological niche. That space could be filled by another invasive mosquito, like the Asian Tiger mosquito, which is equally aggressive and also spreads disease.
The modification strategy avoids this problem. By making Aedes aegypti unable to transmit diseases, we keep the mosquito population stable. The ecological niche stays filled, preventing other species from moving in. We solve the disease problem directly.
The trade-off comes down to this: With elimination, you get rid of bites and disease, but risk replacing one mosquito problem with another. With modification, you keep getting bitten, but you eliminate the disease and avoid the ecological uncertainty. So the real question is, what matters most to your community: stopping the bites entirely, or stopping just the disease?**</font>


## Dictionaries!

Please read Chapter 9 in the _Computing Companion_ before continuing. The first bit of the narrative below is a quick review of those concepts that you will use throughout the rest of this notebook.

We've seen several Python data types so far: numbers, strings, and lists. More recently, we've also introduced a new type called **dictionaries**.

When we think of a dictionary, we usually think of <font color="green">words</font> and their <font color="maroon">definitions</font> as in:
  * <font color="green">nutella</font>: <font color="maroon">A delicious food</font>
  * <font color="green">snack</font>: <font color="maroon">A small meal</font>
  * <font color="green">SCI10</font>: <font color="maroon">A class at CMC</font>

The code cell below defines a dictionary in Python. Notice the curly ➰
 braces `{` and `}`. These curly braces indicate that this is a dictionary (and not a list, string, or anything else).

Take a look 🔬 at the structure of what's inside the dictionary. After each word there is a colon, followed by the definition of that word. A comma separates each word and its definition.

The _words_ are called **keys** in the dictionary and the _definitions_ are called **values**.

In [1]:
def demo1():
  myDictionary = {"nutella" : "A delicious food", "snack" : "A small meal", "SCI10" : "A class at CMC"}
  print(myDictionary)
  print(len(myDictionary))

demo1()

{'nutella': 'A delicious food', 'snack': 'A small meal', 'SCI10': 'A class at CMC'}
3


The example below shows that we can use `in` to determine if a **key** is defined in the dictionary.

In [2]:
def demo2():
  myDictionary = {"nutella" : "A delicious food", "snack" : "A small meal", "SCI10" : "A class at CMC"}
  if "nutella" in myDictionary:
    print("nutella is in the dictionary!")
  else:
    print("No nutella in the dictionary")
  if "toblerone" in myDictionary:
    print("toblerone is in the dictionary!")
  else:
    print("No toblerone in the dictionary")

demo2()

nutella is in the dictionary!
No toblerone in the dictionary


For each **key**, we can look up its **value**. Notice the use of square brackets `[` and `]` here. That can be challenging to remember because Python uses `{` and `}` when we define a dictionary, but we'll get used to it!

Uh, oh! No "banana" 🍌 in the dictionary! Take a look at Python's error message. `KeyError` means that there is no "banana" **key** in the dictionary.

In [3]:
def demo3():
  myDictionary = {"nutella" : "A delicious food", "snack" : "A small meal", "SCI10" : "A class at CMC"}

  print("For key nutella, the value is", myDictionary["nutella"])
  print("For key snack, the value is", myDictionary["snack"])
  print("For key SCI10, the value is", myDictionary["SCI10"])
  print("For key banana, the value is", myDictionary["banana"])

demo3()

For key nutella, the value is A delicious food
For key snack, the value is A small meal
For key SCI10, the value is A class at CMC


KeyError: 'banana'

### Keys and values don't have to be strings

Keys and values don't have to be strings! In the code cell below, the **keys** are strings (some of SCI 10 professors' names) and the values are **numbers** (the mean number of cups of coffee that they drink each day). Yikes, Prof. Emily!!! ☕

Notice how the for loop is iterating over the keys in the dictionary. We didn't have to call that variable `key`, we could have called it `profName` or anything else! Similarly, the box named `value` could have been called `cupsOfCoffee`. We just use the variable names `key` and `value` here as a reminder of the fact that dictionaries have **keys** with associated **values**.

In [4]:
def demo4():
  cupsPerDay = {"Nia" : 3, "Emily" : 42, "Diana" : 2, "Shibu" : 2}
  for key in cupsPerDay:
    value = cupsPerDay[key]
    print("Prof", key, "drinks", value, "cups per day")

demo4()

Prof Nia drinks 3 cups per day
Prof Emily drinks 42 cups per day
Prof Diana drinks 2 cups per day
Prof Shibu drinks 2 cups per day


**Part 3**

We've seen a function like this before. Take a look at the function and write a contract for what it does. Remember that a contract should specify the (1) type of the input (e.g., "Takes a list of list of numbers as input" or "Takes a list of strings as input", etc.) and (2) the type _and_ meaning of the output (e.g., "Returns a number representing the length of the input list" or "Returns a dictionary in which the keys are the strings in the input list and the values are the lengths of those keys", etc.)

In [5]:
# Takes in a list and returns a dictionary with the word and how many times it appears in the list.
def wordCount(wordList):
  wordCountDict = {}  # This is an empty dictionary
  for word in wordList:
    if word not in wordCountDict:
      wordCountDict[word] = 1
    else:
      wordCountDict[word] = wordCountDict[word] + 1
  return wordCountDict

testList = ["nutella", "coffee", "nutella", "tea", "tea"]
print(wordCount(wordList = testList))

{'nutella': 2, 'coffee': 1, 'tea': 2}


**Part 4**

Here's another version of the same function! Notice that the only change here is that the for loop has changed to iterate over _positions_ (aka _indices_) in the `wordList`. You'll need to fix line 4.

(Why did we make this change? It will be useful to us a little later on when we write Chat-CMC and need positional information in the list!)

In [3]:
def wordCount(wordList):
  wordCountDict = {}  # This is an empty dictionary
  for i in list(range(len(wordList))):
    word = wordList[i]
    if word not in wordCountDict:
      wordCountDict[word] = 1
    else:
      wordCountDict[word] = wordCountDict[word] + 1
  return wordCountDict

testList = ["nutella", "coffee", "nutella", "tea", "tea"]
print(wordCount(wordList = testList))

{'nutella': 2, 'coffee': 1, 'tea': 2}


### DictioNAries meet DNA

Below we represent the codon table that you saw when we talked about translation (of codons to amino acids) - now as a dictionary in Python.

Notice how when we compute the `len` of the `codonDict`, we get the number of keys in the dictionary. Since there are 4 types of bases that can be in 3 different positions, there are 64 combinations (4 x 4 x 4 = 64)

In [4]:
codonDict = {"UUU": "Phe", "UUC": "Phe", "UUA": "Leu", "UUG": "Leu",
              "CUU": "Leu", "CUC": "Leu", "CUA": "Leu", "CUG": "Leu",
              "AUU": "Ile", "AUC": "Ile", "AUA": "Ile", "AUG": "Met",
              "GUU": "Val", "GUC": "Val", "GUA": "Val", "GUG": "Val",
              "UCU": "Ser", "UCC": "Ser", "UCA": "Ser", "UCG": "Ser",
              "CCU": "Pro", "CCC": "Pro", "CCA": "Pro", "CCG": "Pro",
              "ACU": "Thr", "ACC": "Thr", "ACA": "Thr", "ACG": "Thr",
              "GCU": "Ala", "GCC": "Ala", "GCA": "Ala", "GCG": "Ala",
              "UAU": "Tyr", "UAC": "Tyr", "UAA": "stop", "UAG": "stop",
              "CAU": "His", "CAC": "His", "CAA": "Gln", "CAG": "Gln",
              "AAU": "Asn", "AAC": "Asn", "AAA": "Lys", "AAG": "Lys",
              "GAU": "Glu", "GAC": "Glu", "GAA": "Asp", "GAG": "Asp",
              "UGU": "Cys", "UGC": "Cys", "UGA": "stop", "UGG": "Trp",
              "CGU": "Arg", "CGC": "Arg", "CGA": "Arg", "CGG": "Arg",
              "AGU": "Ser", "AGC": "Ser", "AGA": "Arg", "AGG": "Arg",
              "GGU": "Gly", "GGC": "Gly", "GGA": "Gly", "GGG": "Gly",}

print(codonDict)
print(len(codonDict))

{'UUU': 'Phe', 'UUC': 'Phe', 'UUA': 'Leu', 'UUG': 'Leu', 'CUU': 'Leu', 'CUC': 'Leu', 'CUA': 'Leu', 'CUG': 'Leu', 'AUU': 'Ile', 'AUC': 'Ile', 'AUA': 'Ile', 'AUG': 'Met', 'GUU': 'Val', 'GUC': 'Val', 'GUA': 'Val', 'GUG': 'Val', 'UCU': 'Ser', 'UCC': 'Ser', 'UCA': 'Ser', 'UCG': 'Ser', 'CCU': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro', 'ACU': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr', 'GCU': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala', 'UAU': 'Tyr', 'UAC': 'Tyr', 'UAA': 'stop', 'UAG': 'stop', 'CAU': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln', 'AAU': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys', 'GAU': 'Glu', 'GAC': 'Glu', 'GAA': 'Asp', 'GAG': 'Asp', 'UGU': 'Cys', 'UGC': 'Cys', 'UGA': 'stop', 'UGG': 'Trp', 'CGU': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg', 'AGU': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg', 'GGU': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'}
64


But there aren't 64 aminos! There are only 20. This is why the codon table is said to be _degenerate_: multiple codons can encode the same amino acid. You can see that clearly by looking at what was printed when you ran the code cell above. For example, both keys `UUU` and `UUC` code for the value `Phe`.

In the code cell below, we are building a new dictionary, called `aminoDict` that has the amino acids as keys and the corresponding _list_ of codons that represent that acid as the values. This is like a reverse translation!

Now when we compute the `len` of the dictionary we find that there are 21 keys in the dictionary; 20 for the amino acids and 1 for the stop codon.

**Note:** We are using the handy feature that things that you've defined in previous code cells can be used in later code cells in the same notebook. We'll do that here and throughout this notebook (and in future notebooks too!). **But...** if it's been awhile since you last ran those prior code cells, Colab may forget about them. So, you will need to go back and run those prior code cells so that you can use them in later cells.

In [5]:
aminoDict = {}

# codonDict was defined in an earlier code cell and we're using it here.
for codon in codonDict:
  amino = codonDict[codon]
  if amino not in aminoDict:
    aminoDict[amino] = [codon]
  else:
    aminoDict[amino] = aminoDict[amino] + [codon]

print(aminoDict)
print("The length of the dictionary is:", len(aminoDict)) # 20 aminos + stop

{'Phe': ['UUU', 'UUC'], 'Leu': ['UUA', 'UUG', 'CUU', 'CUC', 'CUA', 'CUG'], 'Ile': ['AUU', 'AUC', 'AUA'], 'Met': ['AUG'], 'Val': ['GUU', 'GUC', 'GUA', 'GUG'], 'Ser': ['UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC'], 'Pro': ['CCU', 'CCC', 'CCA', 'CCG'], 'Thr': ['ACU', 'ACC', 'ACA', 'ACG'], 'Ala': ['GCU', 'GCC', 'GCA', 'GCG'], 'Tyr': ['UAU', 'UAC'], 'stop': ['UAA', 'UAG', 'UGA'], 'His': ['CAU', 'CAC'], 'Gln': ['CAA', 'CAG'], 'Asn': ['AAU', 'AAC'], 'Lys': ['AAA', 'AAG'], 'Glu': ['GAU', 'GAC'], 'Asp': ['GAA', 'GAG'], 'Cys': ['UGU', 'UGC'], 'Trp': ['UGG'], 'Arg': ['CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], 'Gly': ['GGU', 'GGC', 'GGA', 'GGG']}
The length of the dictionary is: 21


**Part 5**

Complete the code cell below to create a new dictionary called `numCodonsDict` with amino acids as its keys and the number of codons that encode that amino acid as the values. You can use the `aminoDict` that we made in the code cell above to compute the `len`gth of the list that contains all the codons that code for each amino acid. The code cell below should display something that starts off like this:
<pre>
{'Phe': 2, 'Leu': 6, 'Ile': 3, 'Met': 1, ...
</pre>

Notice that the code that you write here will not be inside a function: There is no `def` followed by a function name. That's fine. Although we _normally_ use functions so that we can use them later by name, it's also fine to have code that is not inside a function - it's just code that is run when we run the code cell. We've actually already done this a lot. For example, in the code cell above, notice that lines 11 and 12 are not inside a function - they are just run when we run the code cell.

In [6]:
numCodonsDict = {}
for amino in aminoDict:
  numCodonsDict[amino] = len(aminoDict[amino])

print(numCodonsDict)

{'Phe': 2, 'Leu': 6, 'Ile': 3, 'Met': 1, 'Val': 4, 'Ser': 6, 'Pro': 4, 'Thr': 4, 'Ala': 4, 'Tyr': 2, 'stop': 3, 'His': 2, 'Gln': 2, 'Asn': 2, 'Lys': 2, 'Glu': 2, 'Asp': 2, 'Cys': 2, 'Trp': 1, 'Arg': 6, 'Gly': 4}


**Part 6**


Let's implement the process of translation from mRNA to proteins in Python. Complete the `translate` function below. This function should take a list of codons as input and use the `codonTable` dictionary to look up the amino acid **and return** a string of amino acids (aka, a protein!) with each amino acid separated by a `-`.

For example...
<pre>
translate(["AUG", "UUU", "AGA", "CGC", "UAA"]))
</pre>
... returns the string...
<pre>
"Met-Phe-Arg-Arg-stop"
</pre>



Notice that there is no `-` after the `stop` codon!

In [20]:
def translate(codonList):
  output = ""

  if codonList[0] != "AUG":
    return "Sequence does not begin with start codon."

  # codonList[len(codonList)-1] extracts the codon string in the last location in the list
  elif codonList[len(codonList)-1] not in ["UAA", "UAG", "UGA"]:
    return "Sequence does not end with stop codon."

  else:
    for codon in codonList:
      amino = codonDict[codon]
      if amino == "stop":
        output = output
      else:
        output = output + amino + "-"

    # I'm not sure, is there a better way to do this?
    if output[-1] == "-":
      return output[:-1]
    return output

print(translate(codonList = ["AAG", "UUU", "AGA"])) # Sequence does not begin with start codon.
print(translate(codonList = ["AUG", "UUU", "AGA", "CGC", "ACC"])) # Sequence does not end with stop codon.
print(translate(codonList = ["AUG", "UUU", "AGA", "CGC", "UAA"])) # Met-Phe-Arg-Arg-stop

Sequence does not begin with start codon.
Sequence does not end with stop codon.
Met-Phe-Arg-Arg


## Tuples: First cousins of lists!

Next, let's see another new datatype called _tuples_. They are _nearly_ identical to lists.

In [21]:
def addUp(myTuple):
  counter = 0
  for item in myTuple:
    print("The next item is", item)
    counter = counter + item
  return counter

tupleExample = (10, 20, 42) # A tuple! It looks like a list but uses ( and )
sum = addUp(myTuple = tupleExample)
print(sum)

The next item is 10
The next item is 20
The next item is 42
72


The main difference between tuples and lists are:
  * Tuples use parentheses `(` and `)` rather than brackets `[`, `]`.
  * Tuples _can_ be used as the **keys** in dictionaries while lists cannot. (Any data type can be used as  **values** in dictionaries however, including lists, tuples, numbers, strings, and even other dictionaries!)
  * (There are a few other differences, but they won't arise in this course.)


Though tuples are created by using parentheses `()`, items within the tuple can still be indexed using brackets `[]` with an index inside those brackets. This is similar to how dictionaries are created using curly brackets `{}` but we can extract values associated with keys using brackets `[]`. Here's an example...

In [22]:
myTuple = ("a", 42, ["b"])
print("The first item in myTuple is", myTuple[0])
print("The second item in myTuple is", myTuple[1])
print("The third item in myTuple is", myTuple[2])

The first item in myTuple is a
The second item in myTuple is 42
The third item in myTuple is ['b']


In [23]:
myDict = {"a": "First letter", "b": "Second letter"}
print("The value for the key 'a' is", myDict["a"])
print("The value for the key 'b' is", myDict["b"])

The value for the key 'a' is First letter
The value for the key 'b' is Second letter


### Tuples and lists as keys in dictionaries

Imagine that we'd like to have a dictionary where the keys are pairs, like first name and last name and the values are the lists of foods that this person likes.

Try that below, but Python complains...

In [24]:
key1 = ["Colin", "Rathbun"]
value1 = ["cookies", "almonds", "nutella"]
key2 = ["Nia", "Walker"]
value2 = ["sushi", "hazelnuts"]
# We start with an empty dictionary...
foodLikes = {}
# ... and try to add key-value pairs
foodLikes[key1] = value1
foodLikes[key2] = value2
print(foodLikes)

TypeError: unhashable type: 'list'

The error message is telling us that the dictionary doesn't like a list as a key. But, by using a tuple instead, everything is fine. Notice that the values can still be lists!

In [25]:
key1 = ("Colin", "Rathbun")
value1 = ["cookies", "almonds", "nutella"]
key2 = ("Nia", "Walker")
value2 = ["sushi", "hazelnuts"]
# We start with an empty dictionary...
foodLikes = {}
# ... and try to add key-value pairs
foodLikes[key1] = value1
foodLikes[key2] = value2
print(foodLikes)

{('Colin', 'Rathbun'): ['cookies', 'almonds', 'nutella'], ('Nia', 'Walker'): ['sushi', 'hazelnuts']}


## For loops on for loops! (aka "Nested 🪆 for loops")

We've already introduced the idea of nested for loops - one for loop within another for loop.

Let's start with a single for loop to remind ourselves of a key concept. In the function in the code cell below, notice that the for loop in line 3 has a block of code that "belongs to it" - lines 4 through 9 are indented underneath the for loop. That means that for each `num` in `numList`, Python will perform the instructions in that nested block.

In [26]:
def simple(numList):
  print("Getting started")
  for num in numList:
    print("  The next number is", num)
    if num > 42:
      print("  Too large!")
    elif num < 42:
      print("  Too small!")
    else:
      print("  Just right!")
  print("All done")

simple(numList = [5, 100, 42, 7])


Getting started
  The next number is 5
  Too small!
  The next number is 100
  Too large!
  The next number is 42
  Just right!
  The next number is 7
  Too small!
All done


Next, consider the function below. Notice that the for loop in line 3 has a block of code that "belongs to it" - that's the block indented below it in lines 4 and 5. So, by the same reasoning as above, for each `base` in `baseList`, Python must complete lines 4 and 5 before coming back to line 3 to consider the next `base`.

In this example, the first `base` is `"thin crust"`. So, now, Python must complete the block of code in lines 4 and 5 before coming back to the next `base`.

To do that, Python starts at line 4. Here, it must loop through each `topping` in `toppingsList`. It goes through them one-by-one. When it's all done going through the toppings, it has fulfilled its obligation. Now, the block of code in lines 4 and 5 is complete and we come back to line 3 to consider the next `base`.

It's important to note that nested for loops are not _magic_. Just like we can have if-elif-else statements inside the block of a for loop, we can have another for loop inside that block.

In [27]:
def combos(baseList, toppingsList):
  print("Available combinations:")
  for base in baseList:
    for topping in toppingsList:
      print(base, "with", topping)
  print("Ciao!")

combos(baseList = ["thin crust", "thick crust"], toppingsList = ["mushrooms", "basil", "tomato"])

Available combinations:
thin crust with mushrooms
thin crust with basil
thin crust with tomato
thick crust with mushrooms
thick crust with basil
thick crust with tomato
Ciao!


Next, take a look at a variant of the `combos` function below. Notice that we have a `print` statement in line 4 to print the current `base`. Then, at line 5 we have a second `for` loop for printing all of the toppings. As was the case above, the `base` does not change until the second `for` loop has completed going through the list of `topping`s.

In [28]:
def combos1(baseList, toppingsList):
  print("Available combinations")
  for base in baseList:
    print("Base", base)
    for topping in toppingsList:
      print(" topping", topping)
  print("Ciao!")

combos1(baseList = ["thin crust", "thick crust"], toppingsList = ["mushrooms", "basil", "tomato"])


Available combinations
Base thin crust
 topping mushrooms
 topping basil
 topping tomato
Base thick crust
 topping mushrooms
 topping basil
 topping tomato
Ciao!


**Part 7**

The code cell below defines a function that takes a list of DNA sequences (strings) as input and prints all pairs of DNA sequences that have the same length. For the `testList` test input, this function should print:
<pre>
ATC ATC
ATC TTT
TTAC TTAC
TTAC GGGA
GGGA TTAC
GGGA GGGA
TTT ATC
TTT TTT
</pre>

Fill in the missing code. Recall that `len` is a built-in function that takes a string or list as input and returns its length.

In [29]:
def sameLength(DNAstrings):
  for DNA1 in DNAstrings:
    for DNA2 in DNAstrings:
      if len(DNA1) == len(DNA2):
        print(DNA1 + " " + DNA2)

testList = ["ATC", "TTAC", "GGGA", "TTT"]
sameLength(DNAstrings = testList)

ATC ATC
ATC TTT
TTAC TTAC
TTAC GGGA
GGGA TTAC
GGGA GGGA
TTT ATC
TTT TTT


**Part 8**

Next, in the code below, write a new version of `sameLength` called `sameLength1` that **returns** the list of all tuples of two DNA strings that have the same length, _except_ that it should not include a DNA sequence and itself.

For the provided `testList`, line 9 should print
<pre>
[('ATC', 'TTT'), ('TTAC', 'GGGA'), ('GGGA', 'TTAC'), ('TTT', 'ATC')]
</pre>

In [34]:
def sameLength1(DNAstrings):
  output = []
  for DNA1 in testList:
    for DNA2 in testList:
      if DNA1 == DNA2:
        output.append((DNA1, DNA2))

  return output


testList = ["ATC", "TTAC", "GGGA", "TTT"]
result = sameLength1(DNAstrings = testList)
print(result)

[('ATC', 'ATC'), ('TTAC', 'TTAC'), ('GGGA', 'GGGA'), ('TTT', 'TTT')]


## Reading text files

Next, copy the `GreenEggsAndHam.txt` and `ShakespeareSonnets.txt` files from the [Week 8 folder](https://drive.google.com/drive/folders/1yxAg4HiVFdJdal6mT9DgchkaFj5HGFHF?usp=drive_link) into *your own* `SCI10 > Week8` folder on Google Drive. Please be sure that you do not change the file names! (Note that we are doing HW7 but it is `Week8` of the course. Be sure that your `SCI10` folder is at the top level of your google drive rather than inside some other folder. Be careful about spelling and spaces. The code below is going to look for the files in your `SCI10` folder and, within that folder, in the `Week8` subfolder.)

Then, copy the `wordCount` function from above (Part 3) to the cell below.

In [38]:
from google.colab import drive
drive.mount('/content/drive')

# This function takes in a filename (in Google Drive)
# and returns the list of words in that file.
def readFile(filename):
  with open(filename, "r") as file:
    fileContents = file.read()
    wordList = fileContents.split()
    return wordList

# Paste your wordCount function here. Use the version from the code cell
# above.

def wordCount(wordList):
  wordCountDict = {}  # This is an empty dictionary
  for word in wordList:
    if word not in wordCountDict:
      wordCountDict[word] = 1
    else:
      wordCountDict[word] = wordCountDict[word] + 1
  return wordCountDict

greenList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/GreenEggsAndHam.txt")
sonnetList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/ShakespeareSonnets.txt")
print("GreenEggsAndHam.txt has", len(greenList), "words")
print(greenList)
print()
print("ShakespeareSonnets.txt has", len(sonnetList), "words")
print(sonnetList)
print()
countGreen = wordCount(wordList = greenList)
print(countGreen["ham."]) # should be 5
countSonnet = wordCount(wordList = sonnetList)
print(countSonnet["we"])  # should be 13

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
GreenEggsAndHam.txt has 777 words
['That', 'Sam-I-am!', 'That', 'Sam-I-am!', 'I', 'do', 'not', 'like', 'that', 'Sam-I-am!', 'Do', 'you', 'like', 'green', 'eggs', 'and', 'ham?', 'I', 'do', 'not', 'like', 'them,', 'Sam-I-am.', 'I', 'do', 'not', 'like', 'green', 'eggs', 'and', 'ham.', 'Would', 'you', 'like', 'them', 'here', 'or', 'there?', 'I', 'would', 'not', 'like', 'them', 'here', 'or', 'there.', 'I', 'would', 'not', 'like', 'them', 'anywhere.', 'I', 'do', 'not', 'like', 'green', 'eggs', 'and', 'ham.', 'I', 'do', 'not', 'like', 'them,', 'Sam-I-am.', 'Would', 'you', 'like', 'them', 'in', 'a', 'house?', 'Would', 'you', 'like', 'them', 'with', 'a', 'mouse?', 'I', 'do', 'not', 'like', 'them', 'in', 'a', 'house.', 'I', 'do', 'not', 'like', 'them', 'with', 'a', 'mouse.', 'I', 'do', 'not', 'like', 'them', 'here', 'or', 'there.', 'I', 'do', 'not', 'like', 'them', 'an

In [42]:
print(countSonnet["love"])
print(countGreen["eggs"])
#print(countGreen["ham"])
print(countGreen["ham."])

for word in countGreen:
  if countGreen[word] == 42:
    print(word)

for word in countSonnet:
  if countSonnet[word] == 42:
    print(word)


103
11
5
thee,


**Part 9** (10 points)

Using the above files, the `wordCount` function from Part 3, and any additional code that you need to write (you can start a new code cell to do that), answer the five questions below.


1. How many times does <tt>love</tt> appear in _Shakespeare's Sonnets_ ?
+ Answer: 103

2. How many times does <tt>eggs</tt> appear in _Green Eggs and Ham_ ?
+ Answer: 11

3. How many times does <tt>ham</tt> appear in _Green Eggs and Ham_ ?
+ Answer: 0

4. How many times does <tt>ham.</tt> (with a period) appear in _Green Eggs and Ham_ ?
+ Answer: 5

5. What is a word that appears **exactly 42 times** in one of these files ? (Share the word and the file it was from!) **Note:** This will require writing a few lines of new code, possibly involving a for loop and a conditional statement.
+ Answer: thee,

## Chat-CMC!

Now, we'll use dictionaries to implement a small version of a generative AI tool totally from scratch!

Rather than a large language model like Chat-GPT that trains on "everything out there", our Chat-CMC system will be trained on text files like _Green Eggs and Ham_ or Shakespeare's Sonnets.

To that end, we'll need a function called `learn(wordList)`
that takes a list of words from a text file
as input and returns a dictionary where each **key** is a word in `wordList` and the value for each key is the _list_ of words that come immediately after that key.

For example, suppose that our text file was Prof. Lars' new (still unpublished) _very short novel_  called "Spam and Poptarts". Here is the content of that novel as a list of words called `wordList`:

```wordList = ["I", "like", "spam.", "I", "love", "poptarts!"]```

The `learn` function should create and return the following dictionary:

```
   {'I': ['like', 'love'],
    'like': ['spam.'],
    'spam.': ['I'],
    'love': ['poptarts!']}
```

Notice that after `"I"` we find `"like"` and we find `"love"`. So, for the **key** `"I"` we have the **value** `['like', 'love']`.

Similarly, after the `"like"` we find `"spam."`. So, for the **key** `"like"` we have the **value** `['spam.']`.

Notice that after `"poptarts!"` we have nothing. For this reason, when we build this dictionary, we look at every word in the word list _except_ for the last word in the word list.





**Part 10** (10 points)

The `learn(wordList)` function will be very similar to the `wordCount` function above! In fact, you might want to copy-and-paste the `wordCount` function into the code cell below so that you can refer to it as you write the `learn` function. It will be _very_ similar.

For the sake of a running example, consider again the sample example
<pre>
wordList = ["I", "like", "spam.", "I", "love", "poptarts!"]
</pre>

Here's how the `learn` function will work.
+ Start with an empty dictionary called `nextWordDict`. Ultimately, we'll return that dictionary. Those first and last lines of code are provided. This is the "sandwich" 🥪 approach of writing the first and last lines of code first.
+ Loop over the indices (locations) of the `wordList`, but stop one before the last index. For example, for the `wordList` above, the length is 6, so the last word `"poptarts!"` has index 5, so we'll iterate from indices 0 through 4. The word at index 4 is `"love"` which is the last word to have a word that follows it.
   + Get the word at the current index. Call it `word`. Get the next word at the next index. Call it `nextWord`. (It's a good thing that we're looping over locations/indices! Otherwise, there would be no way to get the next word! The next word is at the index following the current word. Location is everything!)
   + If `word` is **not** yet in the `nextWordDict` dictionary, that is, there is no _key_ for that word yet in the dictionary, then we assign a _value_ to that key. That value is **a list**! What list? (This is _very_ much like what we did in the `wordCount` function, but rather than making the value be a number, we're making it a list!)
   + Otherwise, if `word` is in `nextWordDict`, that is, we've seen this word before, then there is already a value associated with that key. That value is a list, and we just need to add it to the end of the list.



For our running example of...
```
wordList = ["I", "like", "spam.", "I", "love", "poptarts!"]
```
... your code will build up the dictionary like this...
```
nextWordDict is {'I': ['like']}
nextWordDict is {'I': ['like'], 'like': ['spam.']}
nextWordDict is {'I': ['like'], 'like': ['spam.'], 'spam.': ['I']}
nextWordDict is {'I': ['like', 'love'], 'like': ['spam.'], 'spam.': ['I']}
nextWordDict is {'I': ['like', 'love'], 'like': ['spam.'], 'spam.': ['I'], 'love': ['poptarts!']}
```

As another example, if...

<pre>
wordList = ["I", "like", "spam", "and", "I", "like", "poptarts", "and", "I", "love", "nutella!"]
</pre>

... then the value associated with key `"I"` would be the list `["like", "like", "love"]`. Notice that `"like"` appears twice there.

In [43]:
# You might want to copy-and-paste the wordCount function into this cell so that
# you can refer to it since the two functions will be quite similar.

def wordCount(wordList):
  wordCountDict = {}  # This is an empty dictionary
  for word in wordList:
    if word not in wordCountDict:
      wordCountDict[word] = 1
    else:
      wordCountDict[word] = wordCountDict[word] + 1
  return wordCountDict

# Takes input of a list of words, and returns dictionary with each word and words following each word.
def learn(wordList):
  nextWordDict = {}
  for i in range(len(wordList)-1):  # This chops off the last word in the wordList
    word = wordList[i]
    nextWord = wordList[i+1]
    if word not in nextWordDict:
      nextWordDict[word] = [nextWord]
    else:
      nextWordDict[word].append(nextWord)
  return nextWordDict

testList  = ["I", "like", "spam.", "I", "love", "poptarts!"]
print(learn(wordList = testList))

{'I': ['like', 'love'], 'like': ['spam.'], 'spam.': ['I'], 'love': ['poptarts!']}




As before, if all goes well, the test above should print:
<pre>
{'I': ['like', 'love'], 'like': ['spam.'], 'spam.': ['I'], 'love': ['poptarts!']}
</pre>

**Part 11** (10 points)

Now, we're ready to generate text!

Start by copy-and-pasting your `learn` function from the code cell above into the code cell below.

Then, write the `generate` function. Some guidance and details:
+ `generate` will take as input the first word (string) in the text to be generated (we'll call that input `startWord`). It also takes in the dictionary itself - the one learned by the `learn` function, we'll keep calling it `nextWordDict`, and a third input, which is a positive number indicating the number of words to generate. We'll call this `numWords`.

With those three inputs, here's how
`generate(startWord, nextWordDict, numWords)` will work:
* The function maintains a `currentWord` which is the word that it will generate next. This word is initialized to the `startWord`.
* For `numWords` iterations:
  * Print the `currentWord`. If you use Python's built-in `print` function, then each word will be on its own line. But, if you use the provided `nicePrint` function, then multiple words will be on the same line, which is much nicer to read!
  * If the `currentWord` is `in` the dictionary, then the dictionary provides us with the list of possible next words! We choose a next word at random from that list by using `random.choice`. Put that word into a bucket called `nextWord`.
  * Otherwise, if the `currentWord` is not in the dictionary, we set the `nextWord` to be the `startWord`.
  * As with many of our programs, before the loop iteration ends, be sure to make `currentWord = nextWord`. That way, each "child word" becomes the next iteration's "parent word," and so on!

Write and test your `generate` function in the code cell below. It should be about 6-8 lines of code.

Notice that the generator may stop generating words in "mid-sentence" because we've specified exactly how many words to generate. This is OK -- we can ask it to generate more!

In [55]:
import random
COLUMN = 0      # Global variable used for nice printing
NUMCOLUMNS = 50
random.seed(46)

# Prints nicely!
def nicePrint(string):
  global COLUMN
  if COLUMN + len(string) > NUMCOLUMNS:
    print(string) ; COLUMN = 0
  else:
    print(string, end=" ") ; COLUMN = COLUMN + len(string) + 1

# Copy your learn function here along with its contract.
# Takes input of a list of words, and returns dictionary with each word and words following each word.
def learn(wordList):
  nextWordDict = {}
  for i in range(len(wordList)-1):  # This chops off the last word in the wordList
    word = wordList[i]
    nextWord = wordList[i+1]
    if word not in nextWordDict:
      nextWordDict[word] = [nextWord]
    else:
      nextWordDict[word].append(nextWord)
  return nextWordDict

# Generates numWords of text starting with startWord. Uses the
# nextWordDict to get the list of possible next words and then
# chooses one of those words using random.choice.
def generate(startWord, nextWordDict, numWords):
  currentWord = startWord   # getting started with currentWord
  nicePrint(currentWord)
  for _ in range(numWords):
    nextWord = random.choice(nextWordDict[currentWord])
    nicePrint(nextWord)
    currentWord = nextWord


testList = ["I", "like", "spam.", "I", "love", "love", "poptarts!", "love", "is", "sweet", "like", "poptarts!"]

nextWordDict = learn(wordList = testList)  # learn!
generate(startWord = "I", nextWordDict = nextWordDict, numWords = 20)  # generate!

I like poptarts! love is sweet like spam. I like poptarts!
love love love is sweet like spam. I love poptarts!


If all goes according to plan, you should get the following output when using a random seed of 46 (and our provided `nicePrint` function):
<pre>
I like poptarts! love is sweet like spam. I like poptarts!
love love love is sweet like spam. I love
</pre>

**Part 12**

Now, let's train-and-generate from our data files! Copy your `learn` and `generate` functions in the cell below. Try it out on both text files. (You can choose which file you want to use by changing which line gets commented in or out when reading the text file.)

Once your code is working, you can comment out `random.seed(46)` to see different creations!

In [57]:
from google.colab import drive
drive.mount('/content/drive')

import random
NUMCOLUMNS = 50
COLUMN = 0
#random.seed(46)

# This function takes in a filename (in Google Drive)
# and returns the list of words in that file.
def readFile(filename):
  with open(filename, "r") as file:
    fileContents = file.read()
    wordList = fileContents.split()
    return wordList

# Prints nicely!
def nicePrint(string):
  global COLUMN
  if COLUMN + len(string) > NUMCOLUMNS:
    print(string) ; COLUMN = 0
  else:
    print(string, end=" ") ; COLUMN = COLUMN + len(string) + 1

# Paste your learn and generate code here!
# Takes input of a list of words, and returns dictionary with each word and words following each word.
def learn(wordList):
  nextWordDict = {}
  for i in range(len(wordList)-1):  # This chops off the last word in the wordList
    word = wordList[i]
    nextWord = wordList[i+1]
    if word not in nextWordDict:
      nextWordDict[word] = [nextWord]
    else:
      nextWordDict[word].append(nextWord)
  return nextWordDict

# Generates numWords of text starting with startWord. Uses the
# nextWordDict to get the list of possible next words and then
# chooses one of those words using random.choice.
def generate(startWord, nextWordDict, numWords):
  currentWord = startWord   # getting started with currentWord
  nicePrint(currentWord)
  for _ in range(numWords):
    nextWord = random.choice(nextWordDict[currentWord])
    nicePrint(nextWord)
    currentWord = nextWord


wordList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/GreenEggsAndHam.txt")
#wordList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/ShakespeareSonnets.txt")
print("Successfully read in the word list!")

nextWordDict = learn(wordList)  # learn!
print("Successfully finished learning; about to start generating!\n")

startPos = random.choice(range(len(wordList)-1))  # index of a random word
startWord = wordList[startPos]  # the random word to start generating from
generate(startWord, nextWordDict, 200)  # generate!

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Successfully read in the word list!
Successfully finished learning; about to start generating!

I will not like them, Sam-I-am. Could you? In a tree!
Not in a fox. Not in a tree. I do not like green eggs
and ham. I could you, would not, in a tree. Not in a
box. And I would not like them in a mouse. I do! I do
not eat them here or there. I do not like them, Sam-I-am.
Would you, could not, could you, on a train! A train!
A train! A train! A train! A train! A train! A train!
Could you, could not like them, Sam-I-am. A train! A
train! Could you, in the dark! Not in a mouse. And in
the rain. And in the dark? I do not like them with a
car. Not on a goat. I do not like them anywhere. I would
not, in the dark. Not in a boat? I will not like green
eggs and ham. I could not, would not, would not, in
the dark? I will eat them in a house. Not in a fox.
I will eat them on

## Chat-CMC-2

One of the shortcomings of Chat-CMC is that the text that it generates is not very good! A better model would use **more words as context** when deciding on the next word to generate.

<font color="#87212e">Chat-CMC</font> used just one word (the current word) for context. For that reason, it's called a "first-order" Markov model.

<font color="#87212e">**Chat-CMC-2**</font> will use _two_ words as context -- it's a "second-order" Markov model. Of course, you could take this further. An optional bonus problem (see below) allows the user to specify how many words to use as context.

For comparison, the current version of ChatGPT purportedly uses a context of somewhere between 4000 and 8000 words. (The details are a trade secret!)
<br>
<br>

**Part 13** (15 points)

Start by copying-and-pasting your previous `learn` and `generate` functions into the code cell below. You will make several modifications to these functions to extend the first-order model (Chat-CMC) to the second-order model (Chat-CMC-2).

Before we talk about any changes to the functions, here's a quick sidebar/reminder about tuples and dictionaries:
* A pair of things in Python is represented as a tuple, which is similar to a list but uses parentheses instead of square brackets:
  * Here is a tuple of a pair of words: `("I", "like")`
  * For comparison, here is a list of a pair of words: `["I", "like"]`
* Dictionaries <font color="LimeGreen">**can** use tuples as keys</font>, but <font color="Red">**can't** use lists as keys</font>. (Lists _can_ be values, but they can't be keys.)

<br/>

**_Changes to_** `learn`

First, you'll need to implement some changes within the `learn` function so that it can learn what follows a given pair of words (rather than what follows a given word):

* The dictionary will now be a dictionary of **pairs** of words. We suggest that you use a name like `wordPair` instead of `word` to remind you of this difference.
* The for loop should no longer iterate through `range(len(wordList)-1)`, but rather `range(len(wordList)-2)`. (Why the change? Try explaining it to your partner!)
* Rather than using a single word as a key, your keys will now be tuples of two words (e.g., `("I", "like")`). Here's a helpful bit of syntax for creating a tuple of two words: `wordPair = (wordList[i], wordList[i+1])`
* Lastly, the nature of the value associated with each key will be the same as before -- it's still a list of (single) words that follow the **two words** of the key. The one change to make here concerns the index of each next word within `wordList`. It should no longer be `i+1`, but rather `...` (you fill in the `...`!).

<br/>

**_Changes to_** `generate`

Next, you'll need to implement some changes within the `generate` function so that it can _use_ two-word tuples to pick the next word to generate!

+ In the names of function arguments, change `startWord` to `startPair`, which will be a tuple of two words (e.g., `("I", "like")`).
+ Before the for loop, set `currentPair` to be `startPair`.
+ Then, in the for loop:
  + Print the _first_ (which index is this?) word of `currentPair`. (Again, we recommend using `nicePrint` for this task.)
  + Next, using `currentPair` as a key:
    + If that tuple is `in` the dictionary, then set the `nextWord` to be one of the _next_ words in the value held by the dictionary `nextWordDict`. Choose one of those at random with `random.choice`. Then, create the next tuple of two words, `nextPair`, using the "old" second word and the new word.
      * To give a concrete example: suppose the values for the `currentPair` of `("I", "like")` is the list `["spam", "spam", "poptarts", "donuts"]`. Suppose the random choice of `nextWord` is `"spam"`. Then, the `nextPair` tuple becomes `("like", "spam")`. Notice that `like` becomes the first word in that tuple and `spam` becomes the second word.
    + If that tuple is **not** in the dictionary, then set `nextPair` to be the `startPair`.
+ Finally, set `currentPair = nextPair` before iterating the loop again.

Write your code in the code cell below and then try it out!

You should get more reasonable looking output than you observed in the original Chat-CMC.

In [58]:
from google.colab import drive
drive.mount('/content/drive')

import random
NUMCOLUMNS = 50
COLUMN = 0
#random.seed(46)

# This function takes in a filename (in Google Drive)
# and returns the list of words in that file.
def readFile(filename):
  with open(filename, "r") as file:
    fileContents = file.read()
    wordList = fileContents.split()
    return wordList

# Prints nicely
def nicePrint(string):
  global COLUMN
  if COLUMN + len(string) > NUMCOLUMNS:
    print(string)
    COLUMN = 0
  else:
    print(string, end=" ")
    COLUMN = COLUMN + len(string) + 1

# Add (and then edit) your learn and generate code here!
# Takes input of a list of words, and returns dictionary with each word and words following each word.
def learn(wordList):
  nextWordDict = {}
  for i in range(len(wordList)-2):  # This chops off the last word in the wordList
    wordPair = (wordList[i], wordList[i+1])
    nextWord = wordList[i+2]
    if wordPair not in nextWordDict:
      nextWordDict[wordPair] = [nextWord]
    else:
      nextWordDict[wordPair].append(nextWord)
  return nextWordDict

# Generates numWords of text starting with startWord. Uses the
# nextWordDict to get the list of possible next words and then
# chooses one of those words using random.choice.
def generate(startPair, nextWordDict, numWords):
  currentPair = startPair   # getting started with currentWord
  for _ in range(numWords):
    nicePrint(currentPair[0])
    if currentPair in nextWordDict:
      nextWord = random.choice(nextWordDict[currentPair])
      nextPair = (currentPair[1], nextWord)
    else:
      nextPair = startPair
    currentPair = nextPair

wordList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/GreenEggsAndHam.txt")
#wordList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/ShakespeareSonnets.txt")
print("Successfully read in the word list!")

nextWordDict = learn(wordList = wordList)  # learn!
print("Successfully finished learning; about to start generating!\n")

startPos = random.choice(range(len(wordList)-2))  # index of a random word
startWords = (wordList[startPos], wordList[startPos+1])  # the random words (a tuple!) to start
generate(startWords, nextWordDict, 200)  # generate!

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Successfully read in the word list!
Successfully finished learning; about to start generating!

a mouse? I do not like them with a mouse. I do not like
them, Sam-I-am. A train! A train! Could you, would you,
with a fox. And I will not eat them in a box. I do not
like them with a fox. I will eat them in a boat. I will
not eat them in a box. I do not like green eggs and
ham! I do not like them, Sam-I-am. Would you? Could
you? In a car? Eat them! Eat them! Eat them! Here they
are. I would not, could not, with a goat! Would you,
could you, in the rain. Not in a tree! I would not,
could not, in the dark! Not in a tree! Not in a tree.
Not in a house. Not in a car! Sam! Let me be! I do not
like them. You will see. You may like them in a house?
Would you like them in a box. I do not like them with
a fox. I will eat them in a car. Not in a house. Not
in a box. I could

## Totally Optional Bonus Problem: <font color="#87212e">**Chat-CMC-_k_**</font>

If you're enjoying this adventure in AI and have a bit of extra time, then consider doing this part.

Write your own code for a model that uses _any_ amount of word context! In this case, your `learn` and `generate` and functions should take as input an additional parameter `k` which is a positive number indicating the context. When `k` is equal to 1, this will be just like our original Chat-CMC. When `k` is equal to 2, this will be just like our Chat-CMC-2. But, when `k` is equal to 3, your new system will consider tuples of 3 words, and so forth.

You will most likely find that the generated text becomes "better" as `k` increases!

### Slicing 🍕

A useful Python concept to solve this problem is **slicing** 🍕. We have talked about indexing before to get items in specific locations from lists or tuples. Slicing takes this one step further. We can use a start index and an end index to get the all the items *within* those two locations. For example:

In [ ]:
myList = ["a", "b", "c", "d", "e", "f", "g", "h"]
startIndex = 1
endIndex = 4
print(myList[startIndex:endIndex])

['b', 'c', 'd']

In [ ]:
myTuple = ("a", "b", "c", "d", "e", "f", "g", "h")
startIndex = 3
endIndex = 7
print(myTuple[startIndex:endIndex])

('d', 'e', 'f', 'g')

Two things to notice here:
* The item at the `endIndex` location is not included in the output. The items *up to* it are.
* We use the brackets `[]` for slicing the tuple even though we used parantheses `()` to initialize it.

We can get more clever with slicing too. For example, if we want to get all the items remaining in a list after a `startIndex` we could use the `len` of the list as our `endIndex`. We could also omit an `endIndex` entirely and still get the same output.


In [ ]:
myList = ["a", "b", "c", "d", "e", "f", "g", "h"]
startIndex = 4
endIndex = len(myList)
print(myList[startIndex:endIndex])
print(myList[startIndex:])

['e', 'f', 'g', 'h']
['e', 'f', 'g', 'h']


Similarly, if we want to get all the items up to an `endIndex` we could use 0 as the `startIndex` or omit it entirely.

In [ ]:
myTuple = ("a", "b", "c", "d", "e", "f", "g", "h")
endIndex = 3
print(myTuple[0:endIndex])
print(myTuple[:endIndex])

('a', 'b', 'c')
('a', 'b', 'c')


Returning back Chat-CMC-k, we can make use of slicing when creating the tuples that encode the context for each word. Take a look at how you used the `wordList` to create the `wordPair`. You could use slicing to do the same thing!

One thing to keep in mind is the change in data type from list to tuple. When we slice things Python keeps the original data type. In the example below, the sliced list is still a list. But we can tell Python to make it a `tuple()` instead too.

In [ ]:
myList = ["a", "b", "c", "d", "e", "f", "g", "h"]
startIndex = 2
endIndex = 5
slicedList = myList[startIndex:endIndex]
print("Sliced list as a list:  ", slicedList)
print("Sliced list as a tuple: ", tuple(slicedList))

Sliced list as a list:   ['c', 'd', 'e']
Sliced list as a tuple:  ('c', 'd', 'e')


With that out of the way, now you can complete the `learn` and `generate` functions below for Chat-CMC-k!

In [ ]:
from google.colab import drive  # Allows us read data file from our google drive
drive.mount('/content/drive')   # Provides access to files in in our drive

import random
NUMCOLUMNS = 50
COLUMN = 0
random.seed(46)

# This function takes in a filename (in Google Drive)
# and returns the list of words in that file.
def readFile(filename):
  with open(filename, "r") as file:
    fileContents = file.read()
    wordList = fileContents.split()
    return wordList

# Prints nicely
def nicePrint(string):
  global COLUMN
  if COLUMN + len(string) > NUMCOLUMNS:
    print(string)
    COLUMN = 0
  else:
    print(string, end=" ")
    COLUMN = COLUMN + len(string) + 1

# Add (and then edit) your learn and generate code here!
# Takes a list of words as input and creates a dictionary with each key being a
# unique tuple of *k* words in that list and each corresponding value being the
# word(s) that immediately follow(s) that tuple of words in the list.
def learn(wordList, k):
  nextWordDict = {}                     # This is an empty dictionary
  for i in range(len(wordList)-k):      # This chops off the last k words in the wordList

  # Write your code here.
  # You might want to copy-and-paste the learn function into this cell so that
  # you can refer to it since the two functions will be quite similar.


  return nextWordDict

# Generates numWords of text starting with startTuple. Uses the
# nextWordDict to get the list of possible next words and then
# chooses one of those words using random.choice.
def generate(startTuple, nextWordDict, numWords):
  currentTuple = startTuple    # getting started with currentWord

  # Below, write the code for your generate function
  # You might want to copy-and-paste the generate function from before into this cell so that
  # you can refer to it since the two functions will be quite similar.


wordList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/GreenEggsAndHam.txt")
#wordList = readFile(filename = "/content/drive/My Drive/SCI10/Week8/ShakespeareSonnets.txt")
print("Successfully read in the word list!")

k = 5  # number of words to use as key
nextWordDict = learn(wordList = wordList, k = k)  # learn!
print("Successfully finished learning; about to start generating!\n")

startPos = random.choice(range(len(wordList)-k))  # index of a random word
startWords = tuple(wordList[startPos:startPos+k])  # the random words (a tuple!) to start
generate(startTuple = startWords, nextWordDict = nextWordDict, numWords = 200)  # generate!

## Time to submit!

Congratulations! You've just implemented a small version of a generative AI system! 🤖 Please remember to submit your `HW7.ipynb` notebook on Canvas.